java/compaction-core/src/main/java/sleeper/compaction/job/status/CompactionJobStatusesBuilder.java

```diff
@@ -15,60 +15,98 @@
+        } else {
+            runUpdatesByJobId.computeIfAbsent(jobUpdate.getJobId(), id -> new ArrayList<>())
+                    .add(jobUpdate);
+        }
         return this;
     }
 
     public Stream<CompactionJobStatus> stream() {
-        return createdById.entrySet().stream()
+        return createUpdatesOrderedByUpdateTime().entrySet().stream()
                 .map(entry -> fullStatus(entry.getKey(), entry.getValue()));
     }
 
-    public CompactionJobStatusesBuilder expiryDate(
-            String jobId, Instant expiryDate) {
-        expiryDateById.put(jobId, expiryDate);
-        return this;
-    }
-
     public List<CompactionJobStatus> build() {
         return stream().collect(Collectors.toList());
     }
 
-    private CompactionJobStatus fullStatus(String jobId, CompactionJobCreatedStatus created) {
+    private CompactionJobStatus fullStatus(String jobId, CompactionJobStatusUpdateRecord createdUpdate) {
+        CompactionJobCreatedStatus createdStatus = (CompactionJobCreatedStatus) createdUpdate.getStatusUpdate();
+        List<CompactionJobStatusUpdateRecord> runUpdates = runUpdatesOrderedByUpdateTime(jobId);
+
         return CompactionJobStatus.builder().jobId(jobId)
-                .createdStatus(created)
-                .startedStatus(startedById.get(jobId))
-                .finishedStatus(finishedById.get(jobId))
-                .expiryDate(expiryDateById.get(jobId))
+                .createdStatus(createdStatus)
+                .jobRuns(buildJobRunList(runUpdates))
+                .expiryDate(last(runUpdates)
+                        .map(CompactionJobStatusUpdateRecord::getExpiryDate)
+                        .orElseGet(createdUpdate::getExpiryDate))
                 .build();
     }
+
+    public List<CompactionJobRun> buildJobRunList(List<CompactionJobStatusUpdateRecord> recordList) {
+        Map<String, CompactionJobRun.Builder> taskBuilders = new HashMap<>();
+        List<CompactionJobRun> jobRuns = new ArrayList<>();
+        for (CompactionJobStatusUpdateRecord record : recordList) {
+            String taskId = record.getTaskId();
+            CompactionJobStatusUpdate statusUpdate = record.getStatusUpdate();
+            if (statusUpdate instanceof CompactionJobStartedStatus) {
+                if (taskBuilders.containsKey(taskId)) {
+                    jobRuns.add(taskBuilders.remove(taskId).build());
+                }
+                taskBuilders.computeIfAbsent(taskId, id -> CompactionJobRun.builder())
+                        .startedStatus((CompactionJobStartedStatus) statusUpdate)
+                        .taskId(taskId);
+            } else {
+                jobRuns.add(taskBuilders.remove(taskId)
+                        .finishedStatus((CompactionJobFinishedStatus) statusUpdate)
+                        .taskId(taskId).build());
+            }
+        }
+        taskBuilders.values().stream().map(CompactionJobRun.Builder::build).forEach(jobRuns::add);
```

Comment:
Iterating over the map here will produce a non-deterministic ordering. Can this be tested please? A simple way to do that would be to have two runs on different tasks. I think depending on the task IDs you set and their hash codes, it may come back in the wrong order. I think it should be ordered by the started time of the run.